test/widgets/compose_box_test.dart

```diff
@@ -668,4 +681,81 @@ void main() {
       });
     });
   });
+
+  group('ComposeBox content input scaling', () {
+    const verticalPadding = 8;
+    final stream = eg.stream();
+    final narrow = TopicNarrow(stream.streamId, 'foo');
+
+    Future<void> checkContentInputMaxHeight(WidgetTester tester, {
+      required double maxHeight,
+      required int maxVisibleLines,
+    }) async {
+      final finder = find.byType(ComposeBox);
+
+      double? height;
+      // The content input contains two lines at minimum.
+      String content = 'foo\nfoo';
+
+      // On top of the existing two lines, add exactly the number of lines
+      // needed such that the last line will not visible in the content input
+      // after reaching the max height.
+      for (int i = 0; i < maxVisibleLines - 2 + 1; i++) {
+        // Add one line at a time,
+        // until the content input reaches its max height.
+        await tester.enterText(finder, content);
+        await tester.pump();
+        final newHeight = tester.getRect(contentInputFinder).height;
+        if (newHeight == height) {
+          break;
+        }
+        height = newHeight;
+        content = '$content\nfoo';
+      }
+      check(height).isNotNull().isCloseTo(maxHeight, 0.5);
+      // The last line added did not stretch the content input,
+      // so only the lines before it are at least partially visible.
+      check(content.split('\n').length - 1).equals(maxVisibleLines);
+    }
+
+    testWidgets('normal text scale factor', (tester) async {
+      TypingNotifier.debugEnable = false;
+      addTearDown(TypingNotifier.debugReset);
```

Comment:
The reason this step is needed is the `enterText` calls inside `checkContentInputMaxHeight`, right?

In that case, a good home for this is inside that function. That makes clearer the relationship to why it's needed, and as a bonus makes each test case simpler.